### PR TITLE
feat: generate transcription from recording and API endpoint

### LINKED ISSUE #15347: [CAL-3872] Get Transcripts from Recording
### Is your proposal related to a problem?

Automatically start batch processor job to generate transcript from recording id and API endpoint to fetch the transcripts

<sub>[CAL-3872](https://linear.app/calcom/issue/CAL-3872/get-transcripts-from-recording)</sub>

### PR DESCRIPTION
**Pull request by Udit-takkar**
## What does this PR do?

- Fixes https://github.com/calcom/cal.com/issues/15347
- Fixes CAL-3872 (Linear issue number - should be visible at the bottom of the GitHub issue description)

API Response 

Screenshot 2024-06-06 at 10 42 31 PM

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [x] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [x] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

## How should this be tested?
To test this locally 

1) You can comment this code in apps/web/pages/api/recorded-daily-video.ts

```
  // if (req.headers["x-webhook-signature"] !== computed_signature) {
  //   return res.status(403).json({ message: "Signature does not match" });
  // }
```

2) and make a POST request to /api/recorded-daily-video with body

you can book a meeting locally with daily video and start recording and then get the BookingReference uid from prisma studio and recording id can be fetched from /bookings/past (check network tab) and use booking reference UID as <ROOM_NAME> and recording id as <RECORDING_ID> below

```
{
  "version": "1.1.0",
  "type": "recording.ready-to-download",
  "id": "46-rec-rtd-81db0db6-eb96-4021-a1f6-48c813d542d7",
  "payload": {
    "type": "cloud",
    "recording_id": <RECORDING_ID>,
    "end_ts": 1717688214,
    "room_name": <ROOM_NAME>,
    "start_ts": 1717687980,
    "status": "finished",
    "max_participants": 1,
    "duration": 234,
    "s3_key": "fksfs"
  },
  "event_ts": 1717688213.803
}
```

You should get {message : "success" }

3) then you can make a request to https://api.daily.co/v1/batch-processor?recordingId=9c5ace2c-8c16-416c-9123-b4bf90f4e60d to see if transcription job was submitted.

It takes about 5 - 15 minutes till this batch processor job completes and you can access the transcript download link using the endpoint

### CODE REVIEW

File: apps/web/pages/api/recorded-daily-video.ts [hunk 1]
```diff
@@ -268,6 +271,9 @@ async function handler(req: NextApiRequest, res: NextApiResponse) {
       },
     });
 
+    // Submit Transcription Batch Processor Job
+    await submitBatchProcessorTranscriptionJob(recording_id);
```
Comment (Udit-takkar) [inline apps/web/pages/api/recorded-daily-video.ts:275]: This api endpoint is executed when recording has been stopped and ready to download.

1) submit transcription job 
2) fetch transcription download link from API endpoint (`v1/bookings/[bookingId]/transcripts/[recordingId]`)

File: packages/prisma/zod-utils.ts
Comment (joeauyeung) [inline packages/prisma/zod-utils.ts:1]: The types added to this file feel very specific to this feature from Daily. Can we separate these to a types folder under the daily app package? We should avoid making long files even longer if we don't have to.

File: packages/core/videoClient.ts [hunk 2]
```diff
@@ -301,6 +301,65 @@ const getAllTranscriptsAccessLinkFromRoomName = async (roomName: string) => {
   return videoAdapter?.getAllTranscriptsAccessLinkFromRoomName?.(roomName);
 };
 
+const submitBatchProcessorTranscriptionJob = async (recordingId: string) => {
```
Comment (joeauyeung) [inline packages/core/videoClient.ts:304]: Non-blocking: Something doesn't sit right with me having these Daily specific functions as a part of the video client. Have you thought of any alternatives of where these could live?

File: apps/web/pages/api/recorded-daily-video.ts (hunk 1 above)
Comment (joeauyeung) [reply apps/web/pages/api/recorded-daily-video.ts:275]: For clarification, is the transcription going to be available through the `download link` of the recording, or is it going to be its own separate link that we or the user has to obtain?

**Review by joeauyeung — COMMENTED** [review]
Security and functionality wise I feel comfortable with this PR. 

I can't fully test this PR though. At `getRecordingDownloadLink` I'm getting an unable to access link error. Is there any other way we can test this PR?

**Comment by Udit-takkar** [comment]
> For clarification, is the transcription going to be available through the download link of the recording, or is it going to be its own separate link that we or the user has to obtain?

No. right now there is just v1 API endpoint required by a customer to get the transcription by recording id. in the future I'll also create a tRPC endpoint to use it on our web app.

File: packages/core/videoClient.ts (hunk 2 above)
Comment (Udit-takkar) [reply packages/core/videoClient.ts:304]: Not sure. probably CalVideoClient

File: apps/api/v1/pages/api/bookings/[id]/transcripts/[recordingId]/_get.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  const { id, recordingId } = getTranscriptFromRecordingId.parse(query);
+
+  await checkIfRecordingBelongsToBooking(id, recordingId);
+
+  const transcriptsAccessLinks = await getTranscriptsAccessLinkFromRecordingId(recordingId);
+
+  return transcriptsAccessLinks;
+}
+
+const checkIfRecordingBelongsToBooking = async (bookingId: string, recordingId: string) => {
```
Comment (Udit-takkar) [inline apps/api/v1/pages/api/bookings/[id]/transcripts/[recordingId]/_get.ts:63]: Added another security check

File: apps/web/pages/api/recorded-daily-video.ts (hunk 1 above)
Comment (joeauyeung) [reply apps/web/pages/api/recorded-daily-video.ts:275]: In response to this comment: https://github.com/calcom/cal.com/pull/15358#issuecomment-2154401554

If this throws an error, can we ensure we're catching them properly so it doesn't stop the sending of emails in the next line?

File: apps/api/v1/pages/api/bookings/[id]/transcripts/[recordingId]/_get.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+ *         schema:
+ *           type: integer
+ *         required: true
+ *         description: ID of the booking for which transcripts need to be fetched.
+ *       - in: path
+ *         name: recordingId
+ *         schema:
+ *           type: string
+ *         required: true
+ *         description: ID of the rocording for which transcripts need to be fetched.
```
Comment (joeauyeung) [inline apps/api/v1/pages/api/bookings/[id]/transcripts/[recordingId]/_get.ts:32]: Should also specify in this param the recordingId is coming from Daily

```suggestion
 *         description: ID of the recording for which transcripts need to be fetched.
```

**Review by joeauyeung — CHANGES_REQUESTED** [review]
After the swagger doc is approved ping me and then I'll approve.

**Review by joeauyeung — APPROVED** [review]
Awesome, great job @Udit-takkar!

**Pull request merged**